### ISSUE TITLE
reactNamespace doesn't support `useTranslation` hook

### ISSUE THREAD

**Issue by luisrudge** [i18next/i18next-parser#161]
## What I'm trying to do?

https://github.com/i18next/i18next-parser/blob/46301c7d9cb8e37d832d2259db77b0b01ca365ec/src/transform.js#L250-L258

This code won't evaluate `useTranslation` hooks.

## How am I doing it?

```
const { t } = useTranslation('home');
return <h1>{t(‘welcome’)}</h1>;
```

Here is my config file:

```
module.exports = {
  contextSeparator: '_',
  createOldCatalogs: true,
  defaultNamespace: 't',
  indentation: 2,
  keepRemoved: false,
  keySeparator: false,
  namespaceSeparator: false,
  lexers: {
    js: ['JsxLexer'],
    jsx: ['JsxLexer'],
    default: ['JavascriptLexer']
  },
  lineEnding: 'auto',
  locales: ['en', 'pt-br'],
  output: './src/locales/$LOCALE/$NAMESPACE.json',
  input: undefined,
  reactNamespace: true,
  sort: false,
  useKeysAsDefaultValue: true,
  verbose: false
};
```

The command I'm running:

```
i18next 'src/**/*.js' -c i18next-parser.config.js
```

## What was the result?

`t.json`
```
"welcome”: "t__!NO_KEY_SEPARATOR!__welcome”,
```

## What was I expecting?

`home.json`
```
"welcome”: "home__!NO_KEY_SEPARATOR!__welcome”,
```

**Comment by bitttttten** [comment]
Did you make any progress on this? I am trying to read the source about how it extracts the namespace but I am not having too much luck!

**Comment by bitttttten** [comment]
I have a very crude implementation of this :)

I am using https://www.npmjs.com/package/patch-package and i18next-parser@1.0.0-beta34 for this. Include the following as `patches/i18next-parser+1.0.0-beta34.patch`.

```js
diff --git a/node_modules/i18next-parser/dist/lexers/javascript-lexer.js b/node_modules/i18next-parser/dist/lexers/javascript-lexer.js
index bb00d99..902ff52 100644
--- a/node_modules/i18next-parser/dist/lexers/javascript-lexer.js
+++ b/node_modules/i18next-parser/dist/lexers/javascript-lexer.js
@@ -25,6 +25,8 @@ JavascriptLexer = function (_BaseLexer) {_inherits(JavascriptLexer, _BaseLexer);
 
     _this.acorn = acorn;
     _this.WalkerBase = WalkerBase;
+    _this.currentNamespace = options.defaultNamespace;
+    _this.defaultNamespace = options.defaultNamespace;
 
     // Apply all injectors to the acorn instance
     _this.acornOptions.injectors.reduce(
@@ -51,6 +53,17 @@ JavascriptLexer = function (_BaseLexer) {_inherits(JavascriptLexer, _BaseLexer);
 
     node) {
       var entry = {};
+
+      var isHook = node.callee && node.callee.name === 'useTranslation';
+      if (isHook) {
+        var namespaceKey = node.arguments.shift()
+        if (!namespaceKey.value) {
+          this.currentNamespace = this.defaultNamespace
+        } else if (namespaceKey.type === 'Literal') {
+          this.currentNamespace = namespaceKey.value
+        }
+      }
+
       var isTranslationFunction =
       node.callee && (
       this.functions.includes(node.callee.name) ||
@@ -91,6 +104,10 @@ JavascriptLexer = function (_BaseLexer) {_inherits(JavascriptLexer, _BaseLexer);
             }} catch (err) {_didIteratorError = true;_iteratorError = err;} finally {try {if (!_iteratorNormalCompletion && _iterator.return) {_iterator.return();}} finally {if (_didIteratorError) {throw _iteratorError;}}}
         }
  
+        if (this.currentNamespace) {
+          entry.namespace = this.currentNamespace
+        }
+
         this.keys.push(entry);
       }
     } }, { key: 'concatenateString', value: function concatenateString(

```

**Comment by ermik** [comment]
Using `@next` I have no issues parsing 
```js
<div>{t('namespace:a.bb.ccc')}</div>
``` 
but same usage with backticks:
```js
<div>{t`namespace:a.bb.ccc`}</div>
```
is not parsed.

**Comment by vincaslt** [comment]
Backticks are not supported afaik, that's due to how tagged literals are defined in the AST. Does [i18next](https://www.i18next.com/) support that way of calling it?

**Comment by karellm** [comment]
This is intended. Please read my [answer here](https://github.com/i18next/i18next-parser/issues/83#issuecomment-375907625)

**Comment by timomeh** [comment]
Why was this issue closed? The original issue by @luisrudge was about namespaces inside `useTranslation`, not about backtick support. `useTranslation('home')` still doesn't seem to work.

**Comment by karellm** [comment]
Indeed, my bad. That said it looks like you could use the `functions` option no? Have you tried: ` functions: ['t', 'useTranslation']`?

**Comment by timomeh** [comment]
Setting `lexers.js.functions` to `['t', 'useTranslation']` will just cause `useTranslation` to be recognized as translate-function, which it isn't, and thus will add the namespace of the `useTranslation` hook as translation key.

Input:

```jsx
function SomeThing(props) {
  const { t } = useTranslation('foo')

  return <div>{t('bar')}</div>
}
```

Generated file:

```
// translation.json
{
  "home": "",
  "bar": ""
}
```

Expected:

```
// foo.json
{
  "bar": ""
}
```

**Comment by karellm** [comment]
Got it, thanks for the clarification. PR is welcome!

**Comment by timomeh** [comment]
will take a look at the javascript lexer [tomorrow](https://time.is/de/Cologne)

**Comment by heivo** [comment]
Any progress on this?

**Comment by otaviobps** [comment]
I'm using `i18next-parser 1.0.0-beta38` and couldn't make it work with the `{t}` hook in react.

Here is my config file:
```

// i18next-parser.config.js

module.exports = {
    contextSeparator: "_",
    // Key separator used in your translation keys

    createOldCatalogs: true,
    // Save the \_old files

    defaultNamespace: "main",
    // Default namespace used in your i18next config

    defaultValue: "",
    // Default value to give to empty keys

    indentation: 2,
    // Indentation of the catalog files

    keepRemoved: false,
    // Keep keys from the catalog that are no longer in code

    keySeparator: false,
    // Key separator used in your translation keys
    // If you want to use plain english keys, separators such as `.` and `:` will conflict. You might want to set `keySeparator: false` and `namespaceSeparator: false`. That way, `t('Status: Loading...')` will not think that there are a namespace and three separator dots for instance.

    // see below for more details
    lexers: {
        hbs: ["HandlebarsLexer"],
        handlebars: ["HandlebarsLexer"],
        js: ["JsxLexer"], // if you're writing jsx inside .js files, change this to JsxLexer
        jsx: ["JsxLexer"],
        default: ["JavascriptLexer"]
    },

    lineEnding: "auto",
    // Control the line ending. See options at https://github.com/ryanve/eol

    locales: ["en", "es", "pt"],
    // An array of the locales in your applications

    namespaceSeparator: false,
    // Namespace separator used in your translation keys
    // If you want to use plain english keys, separators such as `.` and `:` will conflict. You might want to set `keySeparator: false` and `namespaceSeparator: false`. That way, `t('Status: Loading...')` will not think that there are a namespace and three separator dots for instance.

    output: "./src/locales/$LOCALE/$NAMESPACE.json",
    // Supports $LOCALE and $NAMESPACE injection
    // Supports JSON (.json) and YAML (.yml) file formats
    // Where to write the locale files relative to process.cwd()

    input: undefined,
    // An array of globs that describe where to look for source files
    // relative to the location of the configuration file

    reactNamespace: true,
    // For react file, extract the defaultNamespace - https://react.i18next.com/components/translate-hoc.html
    // Ignored when parsing a `.jsx` file and namespace is extracted from that file.

    sort: true,
    // Whether or not to sort the catalog

    useKeysAsDefaultValue: true,
    // Whether to use the keys as the default value; ex. "Hello": "Hello", "World": "World"
    // The option `defaultValue` will not work if this is set to true

    verbose: false
    // Display info about the parsing including some stats
};
```

Running `i18next './src/**/*.js' -c src/locales/i18next-parser.config.js` will create just the `main.json` file, while the expected would be the `main.json` file and the others namespaces files. All keys are in this file.

@ermik said that he was using the `@next` (which I am too) and was working fine. Is that any special option I didn't use? My hook usage of `useTranslation` is this one: `const { t } = useTranslation("rate-provider");`

**Comment by karellm** [comment]
Please move out of beta. If this still doesn't work, feel free to open a PR with tests.

**Comment by kirrg001** [comment]
I was running into the same problem. I would be interested in resolving the issue.

**Comment by kirrg001** [comment]
I think I am not sure I 100% understand `reactNamespace` option.

When I set this option to false, the parser will use the namespace defined in `useTranslation("ns")` and creates `ns.json`. Which seems to work fine.

When I set this option to true, it wont. It will create `translation.json`.

The README docs for `reactNamespace` option points to https://react.i18next.com/components/translate-hoc.html, which is a 404? I've never used `withTranslation` HOC or similar.

**Comment by kirrg001** [comment]
I resolved my issue by using `reactNamespace: false`. Please let me know if there is still a need for a fix.

**Comment by karellm** [comment]
@kirrg001 I'm not using any of these options so I'm not of good advice here. It was introduce [here,](https://github.com/i18next/i18next-parser/issues/192) with PRs #195 and #196 

@vincaslt do you rely on these? Could you run the tests?

**Comment by vincaslt** [comment]
Sorry I haven't used namespaces, so not really sure what that option does. From what I gather from here: https://github.com/i18next/i18next-parser/blob/b12e9c479298f8ec435686ea61ccff7fb734624a/src/transform.js#L104 it enables parsing of `translate('namespace')`, I don't understand the point of it, since the namespace declaration in hook seems to be `useTranslation('namespace')`

**Comment by karellm** [comment]
I'm currently running some tests. I'm also not sure about the necessity of this option `reactNamespace` altogether. It looks like it duplicate what is done here: 

https://github.com/i18next/i18next-parser/blob/master/src/lexers/javascript-lexer.js#L105

@luisrudge I'm curious why you set this option to true? You code doesn't seem to rely on that feature at all. Do you need the `translate()` HOC as described in #69 ?

@vincaslt It looks to me we can remove the `reactNamespace` option altogether.

**Comment by Yuripetusko** [comment]
Sorry but what is the status of it?

If I set `reactNamespace: true` then namespaces in `withTranslation` hoc are parsed correctly but not namespaces in `useTranslation ` (instead keys are saved to wrong global default ns)

But if I set `reactNamespace: false` then it's other way around `useTranslation` saves key in correct ns but `withTranslation` doesn't

**Comment by karellm** [comment]
You only should set `reactNamespace` to true if you use the [HOC](https://react.i18next.com/latest/withtranslation-hoc). If not, stay away from that option. 

If you want more help, please provide more code samples or event better a failing test case.

**Comment by Yuripetusko** [comment]
We have mix of class components and function components in our react app, hook cannot be used in class components, so what should I use if I cannot turn them into function components for some reason?

**Comment by karellm** [comment]
I'm not using react nor familiar with it. I don't think I'm the right person to implement this. Maybe @vincaslt can assist you here.

**Comment by Yuripetusko** [comment]
For now I chose to explicitly define namespace with every t function call

**Comment by vincaslt** [comment]
Can I get a small code example that doesn't work right that I could run tests on? From the code I can't see that reactNamespace value has any effect on how namespace is extracted, so it's weird.

**Comment by Yuripetusko** [comment]
Thanks, ok sure will try to make an example today

**Comment by Yuripetusko** [comment]
I looked at code breifly and noticed that you don't even use the latest HOC name when getting namespace? https://github.com/i18next/i18next-parser/blob/a2ec228648c2af2786aed1ceac71a04d3f43e07d/dist/transform.js#L302 this should be 

```
const reactTranslateRegex = new RegExp(
      'withTranslation\\((?:\\s*\\[?\\s*)(' + BaseLexer.stringPattern + ')'
    )
```

Right? I can create a PR if this is indeed a mistake

**Comment by vincaslt** [comment]
I'm not exactly sure of the purpose of that regex. The APIs were probably changing at some point and there are some leftovers and broken stuff. I would expect this to handle your case: https://github.com/i18next/i18next-parser/blob/a2ec228648c2af2786aed1ceac71a04d3f43e07d/src/lexers/javascript-lexer.js#L147

But I would need to make some tests to see what's overriding it or preventing it from working.

**Comment by karellm** [comment]
Fix deployed with `2.2.0`
